### PR TITLE
Add total gradient amplitude transformation

### LINKED ISSUE #470: Total gradient amplitude transformation
**Description of the desired feature:**

A nice transformation to add to `harmonica` is the _total gradient amplitude_ (aka _analytic signal_), defined in Roest et al (1992) as 
$$|A(x, y)| = \sqrt{\left( \frac{\partial M}{\partial x} \right)^2 + \left( \frac{\partial M}{\partial y} \right)^2 + \left( \frac{\partial M}{\partial z} \right)^2}$$

Roest, W. & Verhoef, Joyce & Pilkington, Mark. (1992). Magnetic interpretation using 3-D analytic signal. GEOPHYSICS. 57. 116-125. 10.1190/1.1443174. 

**Are you willing to help implement and maintain this feature?** 

I will try to implement it but no promises about maintenance 😅

### PR DESCRIPTION
**Pull request by leomiquelutti**
Add a new `total_gradient_amplitude()` function that computes the total gradient amplitude for a given 2D grid. It makes use of the derivative functions we already have in Harmonica, computing the horizontal ones through finite differences and the upward one through FFT. Add new gallery and tutorial examples, tests against a synthetic model, and list the new function in the API reference.

**Relevant issues/PRs:**

This PR resolves #470. I

- [x] implemented the _total gradient amplitude_ function in `_transformations.py`
- [x] created the `tga.py` example
- [x] added the section _Total gradient amplitude (aka Analytic Signal)_ to `transformations.rst`
- [x] added `total_gradient_amplitude_test` test to `test_transformations.py`

Notes:
* I had difficulty checking if the mathematical expression in the `transformations.rst` compiled correctly. 
* I also don't know if I commented the function properly since I added a mathematical expression to it.

### CODE REVIEW

**Comment by leomiquelutti** [comment]
Hum I just noticed that I removed the comments of `derivative_northing_kernel` in `filters\_filters.py`. Do I have to undo it or can you refuse only this part of the proposed change?

**Comment by santisoler** [comment]
Thanks for opening this PR @leomiquelutti. I assigned myself to review it. I also started running CI so we it can test and check the style of your PR. Feel free to push fixes if you see any of those failing.

**Comment by leomiquelutti** [comment]
@santisoler I fixed the comments I accidentally removed + ran _make format_ + _check_. It passed some tests that were falling before. Do I have to do something else?

File: doc/user_guide/transformations.rst
Comment (santisoler) [inline doc/user_guide/transformations.rst:401]: I would keep the title to "Total gradient amplitude"
```suggestion
Total gradient amplitude
------------------------
```

We can add a admonition for letting users know that this is a.k.a analytic signal:

```rst
.. hint::

    Total gradient amplitude is also known as *analytic signal*.
```

File: doc/user_guide/transformations.rst
Comment (santisoler) [inline doc/user_guide/transformations.rst:431]: We are not using either of these in the plot, so I think we can just remove them.
```suggestion
```

File: doc/user_guide/transformations.rst
Comment (santisoler) [inline doc/user_guide/transformations.rst:432]: Since the tga is always positive, I think we can go with viridis here and avoid centering the colorscale around zero:
```suggestion
    tmp = tga_grid.plot(cmap="viridis", add_colorbar=False)
```

File: doc/user_guide/transformations.rst
Comment (santisoler) [inline doc/user_guide/transformations.rst:434]: Is this a better title?
```suggestion
    plt.title("Total gradient amplitude of the magnetic anomaly")
```

File: doc/user_guide/transformations.rst
Comment (santisoler) [inline doc/user_guide/transformations.rst:436]: Minor corrections to the units
```suggestion
    plt.colorbar(tmp, label="nT/m")
```

File: examples/transformations/tga.py
Comment (santisoler) [inline examples/transformations/tga.py:9]: Match the underline to the length of the title
```suggestion
Total gradient amplitude of a regular grid
==========================================
```

**Review by santisoler — CHANGES_REQUESTED** [review]
Thanks a lot @leomiquelutti for all these changes, they are greatly appreciated!

I just went through it, this PR is in very good shape. I just added a few comments and suggestions below. Feel free to commit them if you like them, or let me know if you would approach in a different way.

I've also pushed a change adding the new function to `docs/api/index.rst`, so it's listed in the API Reference in the docs.

One thing that we are not documenting properly is that we are not adding new gallery examples, just tutorials. The reason behind it is because we found tutorials way more educational than isolated examples (which need to have a significant plot that can make users relate to them and click them). I think your tutorial is looking awesome, so how would you feel about dropping the example? Let me know what do you think!

File: doc/user_guide/transformations.rst
Comment (santisoler) [inline doc/user_guide/transformations.rst:408]: One other thing I noticed was the difference in the definition of tga between the tutorial and the docstring of the function. Here the absolute value is used, while in the docstring it only appears as $A(x, y)$. I think we should use the same definition in both places. Since the square root is conventionally returning the positive values, I think we can spare the absolute value bars here. What do yo think?

**Comment by santisoler** [comment]
> Actually, there are 4 tests failing:
> 
>     * test_derivative_northing_kernel[1]
> 
>     * test_derivative_northing_kernel[2]
> 
>     * test_derivative_northing_kernel[3]
> 
>     * test_laplace_fft
> 
> 
> Those always failed for me.

Yes, those are failing because of the lines changed in `derivative_northing_kernel`. See one of my comments above.

File: doc/user_guide/transformations.rst
Comment (leomiquelutti) [reply doc/user_guide/transformations.rst:408]: I agree. Probably was a bad copy/paste.

File: harmonica/_transformations.py
Comment (leomiquelutti) [reply harmonica/_transformations.py:349]: Yeap, looks better :)

File: harmonica/filters/_filters.py
Comment (leomiquelutti) [reply harmonica/filters/_filters.py:177]: I am so confused about this. I don't know what I did, but if you solved it is perfectly fine.

**Comment by santisoler** [comment]
> Your corrections and suggestions are all good. I am only confused whether I should resolve the conversations or not. And what else I should do, actually.

Great! I saw you already merged a few of those, awesome!

I don't think there's a correct etiquette for marking conversations as resolved. When I usually go through someone else's review on my code, I mark them as resolve as I apply the suggested changes without any particular comment. If I took another approach, disagree with the suggestion or ask for more clarification I leave them as unresolved. Basically I used the resolve/unresolved state as a reminder of what things need to be done before merging. But that's up to you.

So, for a PR to be merged, it needs to pass all checks in CI, along with an approved review from one of the developers or maintainers. In this case, I'm seeing some of the checks failing, and we still have those modified lines in `derivative_northing_kernel` that should be restored (otherwise the tests on that function will keep failing). I also marked a few conversations as unresolved, since I think they need to be addressed.

If you want, I can help you restore those lines in `derivative_northing_kernel`, no worries.

File: harmonica/filters/_filters.py
Comment (santisoler) [reply harmonica/filters/_filters.py:177]: I haven't solved it yet. But I can do so if you want me to help you with that.

File: doc/user_guide/transformations.rst
Comment (santisoler) [reply doc/user_guide/transformations.rst]: Thanks for committing this suggestion. The second hint wasn't part of it, so it wasn't included in that commit. I just pushed it though, so now you can mark this as resolved.

**Comment by leomiquelutti** [comment]
> I don't think there's a correct etiquette for marking conversations as resolved. When I usually go through someone else's review on my code, I mark them as resolve as I apply the suggested changes without any particular comment. If I took another approach, disagree with the suggestion or ask for more clarification I leave them as unresolved. Basically I used the resolve/unresolved state as a reminder of what things need to be done before merging. But that's up to you.

Thanks!

> If you want, I can help you restore those lines in `derivative_northing_kernel`, no worries.

Yes, please! I gladly accept your help and appreciate it a lot 😃 I thought I had solved it but apparently I didn't.

I'll go and take a look at the unresolved comments to address them.

**Comment by leomiquelutti** [comment]
> So, for a PR to be merged, it needs to pass all checks in CI, along with an approved review from one of the developers or maintainers. In this case, I'm seeing some of the checks failing, and we still have those modified lines in `derivative_northing_kernel` that should be restored (otherwise the tests on that function will keep failing). I also marked a few conversations as unresolved, since I think they need to be addressed.

Do the checks in CI run automatically? Or could I run it before I commit a change? Or can I run it somehow?

File: doc/user_guide/transformations.rst
Comment (leomiquelutti) [reply doc/user_guide/transformations.rst]: And now I solved it (I think)

File: harmonica/_transformations.py
Comment (leomiquelutti) [reply harmonica/_transformations.py]: I implemented this way now.

File: harmonica/filters/_filters.py
Comment (leomiquelutti) [reply harmonica/filters/_filters.py]: I believe that now I solved it.

**Review by leomiquelutti — COMMENTED** [review]
Well, I hope I did it.

**Comment by santisoler** [comment]
> > So, for a PR to be merged, it needs to pass all checks in CI, along with an approved review from one of the developers or maintainers. In this case, I'm seeing some of the checks failing, and we still have those modified lines in `derivative_northing_kernel` that should be restored (otherwise the tests on that function will keep failing). I also marked a few conversations as unresolved, since I think they need to be addressed.
> 
> Do the checks in CI run automatically? Or could I run it before I commit a change? Or can I run it somehow?

They usually run automatically after you push a change. But I think GitHub has strengthen their security and maybe because you're a new contributor I need to approve the run of CI manually. This is annoying... but after we merge this PR, you won't face this issue again.

**Comment by santisoler** [comment]
Indeed, we have the "Require approval for first-time contributors" active in the configuration of GitHub Actions. Here are some [docs about it](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories).

I thought that once I approve the first run, then every run will be approved. But apparently not, I need to go and manually approve every run.

File: harmonica/filters/_utils.py
Comment (santisoler) [inline harmonica/filters/_utils.py:77]:
```suggestion
        If the passed grid is not 2D or if it contains nan values.
```

**Review by santisoler — APPROVED** [review]
@leomiquelutti I just pushed a few minor changes. I saw that you added some checks to the new function that raise errors when the passed grid is not valid (not 2D or if they contain nans). Thanks for that! I just took the lines from `apply_filter` that run those checks and turned them into their own function. Now the `total_gradient_amplitude` function uses that function instead of repeating code. I also added a few tests to check if the function raises errors on an invalid grid, basically to achieve higher test coverage. And a minor adjustment to the LaTeX equation in the user guide.

This is looking great! If CI passes, I think this is ready to be merged.

**Comment by leomiquelutti** [comment]
Hi @santisoler is there anything else I should do about this PR? :) 

I am only waiting for this PR to be done to open a new one to implement the tilt (if you mantainers agree to it, of course).

**Comment by leomiquelutti** [comment]
> That would be great! Is there a reason why you were waiting for this one to be finished? In most cases you can work on several features in parallel (unless they need a particular bit of code that is being developed in another PR). There's no particular rush to have these two transformations, but I was just curious about this comment.

This was my first experience contributing to an open-source project, so I wanted to see that everything went well before starting a new task. 

A question @santisoler: should I keep working in this same branch or create a new one?

**Comment by santisoler** [comment]
> This was my first experience contributing to an open-source project, so I wanted to see that everything went well before starting a new task.

Right, that's a very good reason! Makes total sense.

> A question @santisoler: should I keep working in this same branch or create a new one?

It's preferred if you create a new one. The process would be:

1. Switch to `main` in your local repo.
2. Pull changes from the `main` branch in `fatiando/harmonica`.
3. Create a new branch where you'll start working on the new feature.
4. Open the PR for merging that new branch into `main`.

We do this for every PR, specially when you're working on different things in parallel, you don't want to mix changes from different PRs in a same branch. For example, you can now create one branch for the tilt implementation, and another branch to add yourself tot he `AUTHORS.md` file 🙂 .

**Pull request merged**
